### ISSUE TITLE
✔️ 🚀 Feature Request: Add rel="preload" to fonts

### ISSUE THREAD

**Issue by pensierinmusica** [elementor/elementor#10886]
Hi,

Google Page Speed Insights complains about the fact that the fonts used for the FontAwesome icons don't have the `rel="preload"` attribute.

<img width="1405" alt="Screenshot 2020-03-18 at 17 43 24" src="https://user-images.githubusercontent.com/3594037/76985257-3d91a500-6940-11ea-943e-44b08dab2c04.png">

In my case, the affected files are:

- `/wp-content/plugins/elementor/assets/lib/font-awesome/webfonts/fa-brands-400.woff2`
- `/wp-content/plugins/elementor/assets/lib/font-awesome/webfonts/fa-solid-900.woff2`

And this seems to be where the problem originates in your source code:

- https://github.com/elementor/elementor/blob/bebe98577f3c7c47d7e3de4e534a06a6df96e4c8/assets/lib/font-awesome/css/brands.css#L10 (line 10 and 11)
- https://github.com/elementor/elementor/blob/bebe98577f3c7c47d7e3de4e534a06a6df96e4c8/assets/lib/font-awesome/css/solid.css#L10 (line 10 and 11)
- https://github.com/elementor/elementor/blob/bebe98577f3c7c47d7e3de4e534a06a6df96e4c8/assets/lib/font-awesome/css/all.css#L4423 (line 4423, 4424, 4433, 4434, 4444, and 4445)

Can you please get this fixed asap?

Thanks!

**Comment by jazir5** [comment]
https://wordpress.org/plugins/wp-asset-clean-up/

Use the local font preload feature

**Comment by keremnymn** [comment]
Have you been able to solve this issue? Mine is the same and it delays the page load nearly for 2 seconds.

**Comment by arjun1638** [comment]
Same here, it adds up to 4.8s, can anyone suggest any workaround to fix it temporarily?

**Comment by jazir5** [comment]
@arjun1638 

Read my comment above

**Comment by elyazalee** [comment]
+1 for this fix

**Comment by isotropicdesign** [comment]
Asset clean up is your best bet. It's really easy to do, just add the font URLs to the local fonts field. 
Here's a tutorial with screenshots if you need help: [https://isotropic.co/how-to-preload-elementor-fonts/](https://isotropic.co/how-to-preload-elementor-fonts/)

It's unfortunate that Elementor has not implemented a fix on such a basic issue.

**Comment by heliobentzen** [comment]
Just add font-display: swap in the font-display: swap

**Comment by KonRay** [comment]
+1 for this fix. Really sad that it still exist.

**Comment by dan607** [comment]
I'm doing a lot of projects on Elementor and this issue just killing all my efforts to improve performance. Mobile versions of all websites struggle the most. +5 sec to load time on mobile is massive. 
I would really appreciate Elementor Team to look at this issue as soon as possible.
It shouldn't be a big problem to add rel=preload when loading fonts.

**Comment by oceandiveloper** [comment]
+1 please get this easy fix implemented. thank you

**Comment by montagal** [comment]
++1 Please fix it...

**Comment by Dudaevskiy** [comment]
+1

**Comment by sahilpopli** [comment]
+++++++++++++++1

**Comment by insideth** [comment]
+1, or a way to unload /wp-content/plugins/elementor/assets/lib/font-awesome/webfonts/fa-regular-400.woff2

**Comment by dannydan412** [comment]
+1

**Comment by ZhanArno** [comment]
+1
The same problem.

Elementor team, please fix it.

**Comment by jazir5** [comment]
There's an easy fix for this, just download asset cleanup and add your
fonts to preload in the local fonts section.

Enable remove all google fonts, download your fonts and upload them to
elementors custom fonts section. Then use those fonts throughout your site,
and paste their urls into asset cleanups preload box.

**Comment by marekka** [comment]
+++

**Comment by dannydan412** [comment]
jazir5 please stop replying with this nonsense. That is not an "easy fix", it requires installing a heavily bloated plugin and is an ugly workaround.

**Comment by jazir5** [comment]
You can call it any "ugly work around", but you're wrong. It's an
optimization plugin and will make your site faster. Stop spreading
misinformation. If you don't believe me, install the free version and use
it's asset manager and preload the fonts and tell me your site speed
doesn't improve. Otherwise, please stop talking about things you have no
clue about.

**Comment by dannydan412** [comment]
Preloading fonts doesn't improve speed. I don't think you have any clue how website rendering works. The only issue this resolves is the flash of unstyled content (FOUC) while the fonts haven't been loaded.

I don't want yet another plugin to "make my site faster", there are better, more supported plugins that do that. I've already installed and experimented with this plugin. It has over 100 different features, settings and toggles. I only need 1 of them and it doesn't make sense to install a bloated plugin for 1 feature that should be part of Elementor.
Workarounds are just that - a workaround.

**Comment by GusNinja** [comment]
I solve the problem with a premium plugin WP-Rocket. You only need copy the link that SiteSpeed Insights provides you "/wp-content/plugins/elementor/assets/lib/font-awesome/webfonts/fa-brands-400.woff2" and paste into the WpRocket prechargue section inside fonts section. Easy.

These websites are builded with elementor pro and if you check in google there is not any complain about the slowing typefonts: [https://www.chosqueweb.com]([https://www.chosqueweb.com]), [https://descubreavon.com](https://descubreavon.com)

**Comment by Joni-L** [comment]
I've been looking for other ways around this issue but there doesn't seem to be any filters for fonts and i can't figure out if there's any other frontend outputs that are filterable (in order to solve this via regexps or similar)

**Comment by jazir5** [comment]
Joni, you can use asset cleanup:

https://wordpress.org/plugins/wp-asset-clean-up/

Paste the paths to the fonts in the local fonts section

**Comment by Joni-L** [comment]
Jazir5, we're trying to optimize the installation and remove as much unused stuff as possible. I really don't want to install another plugin just for this functionality. Maybe its worth looking into how "Asset cleanup" does it and strip just that functionality from that plugin.

**Comment by jazir5** [comment]
Then you definitely want asset cleanup. Look into it's js/css manager. It's
explicitly an optimization plugin. It will speed up your site. It has many
other useful features.

**Comment by shaneonabike** [comment]
*sigh* since my ticket has been merged with this one could I ask @pensierinmusica to possible change the title to something more generic to deal with Google fonts. Perhaps: "Provide more flexibility in output of Google fonts"? I have no clue why my ticket aws merged into this one b/c I'm asking for more flexibility to change the payload "prior" to actually writing it out but there you have it..

Here's my original request

**Describe the solution you'd like**
I have three distinct solutions that I think could help this situation.

1. Add an apply_filter for the Google font rendering that actually allows us to modify the added styles. Presently, Elementor adds _all_ the styles no matter whether we use them or not. This can potentially increase the payload and cause slow page speeds . This would also allow us to append a display=swap solution where needed.

A simple solution around https://github.com/elementor/elementor/blob/64137a2af69ad2a9e8411d25d850e2e03078ce67/includes/frontend.php#L789

We could put : `apply_filters('elementor/frontend/google_font_url', $fonts_url);`

2. In the Elementor configuration allow us to select the subset used by default. I realize there is a apply_filter (ref:https://github.com/elementor/elementor/blob/64137a2af69ad2a9e8411d25d850e2e03078ce67/includes/frontend.php#L808) but something more universal that simply allows the user to checkoff the relevant sets would be awesome

**Comment by Joni-L** [comment]
Just to add to @shaneonabike, this issue also applies to Elementor Pro Custom Fonts, not just Google Fonts.

I've not dug deeply enough into the code but i assume it would be possible to add a filter to print_fonts_links() ?

**Comment by vpallas** [comment]
also waiting for this to be fixed properly.
Until then the best option should be to add it to your child theme's functions.php:

`/** Head Font Preloading **/
function font_preloading_preload_key_requests() { ?>
<link rel="preload" as="font" type="font/woff2" href="/wp-content/plugins/elementor/assets/lib/eicons/fonts/eicons.woff2?5.9.1" crossorigin="anonymous">
<link rel="preload" as="font" type="font/woff2" href="/wp-content/plugins/elementor/assets/lib/font-awesome/webfonts/fa-brands-400.woff2" crossorigin="anonymous">
<link rel="preload" as="font" type="font/woff2" href="/wp-content/plugins/elementor/assets/lib/font-awesome/webfonts/fa-brands-900.woff2" crossorigin="anonymous">
<link rel="preload" as="font" type="font/woff2" href="/wp-content/plugins/elementor/assets/lib/font-awesome/webfonts/fa-solid-900.woff2" crossorigin="anonymous">
<?php }
add_action( 'wp_head', 'font_preloading_preload_key_requests' );`

and then clear the cache.

**Comment by ktchan1811** [comment]
Me too, have this issue here.

**Comment by jazir5** [comment]
@dannydan412 

Actually, i'm pretty sure it's you who doesn't know anything about optimization. https://imgur.com/a/62KNQqE. This is my website. Been working on it since the beginning of november. I'm running 51 plugins. I'm running Elementor and Woocommerce. I also have a ton of addons for elementor: powerpack, woolentor, designer powerup, and flexible elementor panel. 

I have a bitcoin payment gateway, a visual woocommerce email template builder, woocommerce subscriptions, woocommerce smart coupons, paypal credit card form on the checkout page, Woocommerce instagram, among many others. I didn't list about 25 plugins, but you get my gist. It also has login with facebook functionality. Cart works too, payment gateways

I challenge you to produce a website that has as many features as mine does, and surpasses my site on speed tests. In fact, I highly, HIGHLY doubt you can produce ANY wordpress website near what mine has with the load times I have been able to achieve. 

To give you some additional context that adds to the hilarity of how much i'm almost certain my website destroys the best website you maintain by a mile, my site is running on a 1 GB ram digital ocean server that i'm managing through cloudways. My site is 810 KB in size, and loads in 1.2 seconds or less consistently, and 800 ms or less on second the second visit, consistently. My TTFB is very low, so is my render blocking score, and content layout shift.

This is my new server: https://imgur.com/a/8qVASJJ. To be honest, i'm still a little disappointed with my gtmetrix score, I can optimize this further. I was able to shave off 140 kb in the time in the couple days since i migrated off the digital ocean server.

I challenge you to show me YOUR actual websites speed tests to compare. I'm so sure that my website destroys yours in performance, that i'll even wager $15 in bitcoin on it. When you can surpass my level of optimization, then you can lecture me on how to improve wordpress performance.

**Comment by dannydan412** [comment]
@jazir5 this "who's is bigger" challenge" belongs on Reddit, not in a professional forum discussing a simple feature request. Nobody cares how fast your website is, it is irrelevant to this discussion. I never made an argument that the plugin you suggested was bad, I simply said you don't need a sledgehammer to crack a nut. Plugins require maintenance. Sometimes a new WP version breaks them. Sometimes they conflict with another plugin. Yes, it's a workaround, but that's all it is - a **workaround**. So what is being asked for here, if it wasn't clear, is to have this built into Elementor, being that it's Elementor that actually injects the fonts into the page to begin with. For a 3rd party plugin to do this it needs to be able to intercept the PHP rendering and modify the DOM. While this is possible, it is not a "good practice". This subtle difference is something that only a developer could understand - the difference between a solution and a hack.

Hope that clears it up.

**Comment by jazir5** [comment]
@dannydan412 Your the one who got argumentative first, so I wouldn't absolve myself of responsibility if I was you. Whether it's a hack or not is irrelevant, what matters is the results. Actual load time is better than following good coding practices and getting a slower loading time. If you can hit all the performance metrics on every speed test like my sites can, the plugins you run can run absolute disasters of spaghetti code. I can't tell you how many times I see people knocking Elementor for being a poorly coded, slow page builder. Or that woocommerce is very heavy and slow. 

If you don't dive deep into optimization, that's true. If you do, it becomes completely irrelevant. The fact of the matter is my sites score 100/100 on lighthouse, and most performance testing sites put mine in the 90th percentile. What matters is the real world user experience. That is the primary reason you even focus on performance, it has a massive impact on your bounce rate.

**Comment by AliHussain85** [comment]
WP-Rocket solves the issue. You can do it so by adding these lines to "Preload Fonts".

`/wp-content/plugins/elementor/assets/lib/font-awesome/webfonts/fa-regular-400.woff2
/wp-content/plugins/elementor/assets/lib/font-awesome/webfonts/fa-solid-900.woff2
/wp-content/plugins/elementor/assets/lib/font-awesome/webfonts/fa-brands-400.woff2`

![image](https://user-images.githubusercontent.com/42221390/103162284-6d93a000-4810-11eb-827a-066e1dd5ff02.png)

**Comment by GuillermoGaete** [comment]
+1

**Comment by batags** [comment]
@vpallas seems like something is missing in your code?

**Comment by bakoline** [comment]
The question is how to disable these fonts to load at all?

/wp-content/plugins/js_composer/assets/lib/bower/font-awesome/webfonts/fa-solid-900.woff2
/wp-content/plugins/js_composer/assets/lib/bower/font-awesome/webfonts/fa-regular-400.woff2
/wp-content/plugins/js_composer/assets/lib/bower/font-awesome/webfonts/fa-brands-400.woff2

These are coming from wp-bakery plugin and pretty sure I don't use them. The total size is more than 160kb...

Any idea will be welcomed

**Comment by hogash** [comment]
This shouldn't be implemented at all. In fact it will hurt your site way more. This tag hints the browser to prioritise the loading of some key assets that should prevent layout shifts, thus improving the Core Vitals.

So preloading content with `rel="preload"` should be only used for essential, above the fold assets, not a gigantic library of font icons. Consider uploading SVG icons ( tens of thousands at http://www.iconfinder.com/ ) to prevent the FA library from loading because most likely most people barely use a few icons from those libraries.

**Comment by warudin** [comment]
+1, super annoying. Especially because the version number of the font needs to be added to the preload link and there's no telling when Elementor will update the font.

**Comment by CyclaKlaus** [comment]
Had this after switching theme, while still having Elementor as builder.  Have SG Optimizer - went and removed the preload, and the error is gone.  Issue for me was font was no longer in use, creating a hang while trying to pre-fetch and not finding.  Much easier than touching any code at all-

**Comment by Michaelgimii** [comment]
Hi @arielk , Will this feature only be added for default fonts, or will custom fonts also benefit from it?

best regards

**Comment by arielk** [comment]
Hi @Michaelgimii 

In recent months we have been working with Google's performance team to improve Elementor's loading.

According to their recommendation, we are going to add the `rel="preconnect"` tag to Google Fonts. DNS, TCP, and TLS happens in parallel with the loading of the Google Fonts CSS file. We effectively improved Google Fonts performance by reducing the load time by ~100 milliseconds.

You can see the code here: https://github.com/elementor/elementor/pull/20251
About `preconnect` tag here: https://www.w3.org/TR/resource-hints/#preconnect 

Thanks

**Comment by Michaelgimii** [comment]
than you @arielk, Good luck.

**Comment by shaneonabike** [comment]
This sounds great thanks for adding that! What about what the initial submitter indicated with regards to FontAwesome which is being loaded locally and doesn't contain these additions?

cc @arielk

**Comment by arielk** [comment]
FontAwesome

> This sounds great thanks for adding that! What about what the initial submitter indicated with regards to FontAwesome which is being loaded locally and doesn't contain these additions?
> 
> cc @arielk

Hi @shaneonabike,

To deal with faster loading of FontAwesome we have already released a new experiment called "Inline Font Icons" (Status: Alpha), in the next releases we are going to release it as a stable, so for now you can just activate this experiment to reduce the loading effect of FontAwesome .

Thanks

**Comment by hogash** [comment]
> This sounds great thanks for adding that! What about what the initial submitter indicated with regards to FontAwesome which is being loaded locally and doesn't contain these additions?
> 
> cc @arielk

Anyone taking performance seriously shouldn't use Font Awesome library at all. Why loading an huge font icon library just for a few icons, when the SVG icon uploader is right next there and there's hundreds of thousands of free icons on sites like iconfinder.com . Also, preloading with rel preload would only things worse.

**Comment by nicholaszein** [comment]
Hello everyone!

We have great news! 🙌 

### 📢 We're happy to announce that the issue you raised was resolved in Elementor **v3.9.0**! 🥳 

Thank you so much for your contribution, but most importantly for your patience! 🙏 

✔️ Please feel free to check it out and update your plugin to the new version!

Cheers 🥂
